landoapi/models/revisions.py

```diff
@@ -32,13 +37,267 @@ class DiffWarningGroup(enum.Enum):
+    )
+    dependent = db.relationship("Revision", uselist=False)
+
+    @classmethod
+    def get_from_revision_id(cls, revision_id):
+        return cls.query.filter(Revision.revision_id == revision_id).one()
+
+    @property
+    def stack_hashes(self):
+        """Return a dictionary with diff and timestamp hashes.
+
+        This property can be used to determine if something changed in the sequence of
+        revisions.
+        """
+        # TODO: possibly add another a status hash, which hashes the sequence of
+        # statuses. In that case, we can be more specific when detecting a change as
+        # some revisions may have an updated timestamp but no meaningful change.
+        stack = [r for r in (self.dependons + self.dependents)]
+        diffs = " ".join([str(r.diff_id) for r in stack]).encode("utf-8")
+        timestamps = " ".join([r.updated_at.isoformat() for r in stack]).encode("utf-8")
+        diffs_hash = hashlib.sha1(diffs).hexdigest()
+        timestamps_hash = hashlib.sha1(timestamps).hexdigest()
+        return {"diffs": diffs_hash, "timestamps": timestamps_hash}
+
+    @property
+    def dependents(self):
+        """Return the current revision and all dependents."""
+        # TODO: rename this to "next sequence".
+        dependents = [self]
+        if not self.dependent:
+            return dependents
+
+        revision = self
+        while revision.dependent:
+            dependents.append(revision.dependent)
+            revision = revision.dependent
+        return dependents
+
+    @property
+    def dependons(self):
+        """Return all revisions that this revision depends on."""
+        # TODO: rename this to "previous sequence".
+        if not self.depends_on:
+            return []
+
+        dependons = []
+        revision = self
+        while revision.depends_on:
+            if revision.depends_on.status == RevisionStatus.LANDED:
+                break
+            dependons.append(revision.depends_on)
+            revision = revision.depends_on
+        dependons.reverse()
+        return dependons
+
+    def get_patch(self):
+        from landoapi.hgexports import build_patch_for_revision
+        from landoapi.workers.revision_worker import call_conduit
+
+        raw_diff = call_conduit("differential.getrawdiff", diffID=self.diff_id)
+        patch_data = self.patch_data or {
+            "author_name": "",
+            "author_email": "",
+            "commit_message": "This is an automated commit message.",
+            "timestamp": int(datetime.now().timestamp()),
+        }
+        return build_patch_for_revision(raw_diff, **patch_data)
+
+    def clear_patch_cache(self):
+        if self.patch_cache_path.exists():
+            self.patch_cache_path.unlink()
+            return True
+        return False
+
+    @property
+    def patch_cache_path(self):
+        from pathlib import Path
+
+        patch_directory = Path("/patches")
```

Comment:
This will be a managed volume when it is deployed to GCP, probably good to keep it consistent with that, so we can just define it in class or module level variable.